app/src/main/java/com/quran/labs/androidquran/ui/PagerActivity.java

```diff
@@ -1837,6 +1887,217 @@ public void onError(@NonNull Throwable e) {
+
+                if (isFirstAyahInSurah) {
+                  startAyahTime = 0;
+                } else {
+                  startAyahTime = startSurahTimingDataArray.get(startAyah);
+                }
+
+                if (isLastAyahInSurah) {
+                  endAyahTime = audioUtils.getSurahDuration(PagerActivity.this,
+                      audioUtils.getSurahAudioPath(audioPathInfo, end.sura));
+                } else {
+                  endAyahTime = startTimeOfAyahAfterEndAyah;
+                }
+
+                boolean startAndEndAyahAreInSameSurah = start.sura == end.sura;
+
+                if (startAndEndAyahAreInSameSurah) {
+                  String audioSegmentPath = audioUtils.getSurahSegment(
+                      audioUtils.getSurahAudioPath(audioPathInfo, start.sura), startAyahTime,
+                      endAyahTime);
+                  audioCacheFilePaths.add(audioSegmentPath);
+                  shareAudioSegment(renameSharableAudioFile(audioSegmentPath));
+                } else {
+                  ArrayList<String> segmentPaths = new ArrayList<>();
+                  int endOfSurah = -1;
+                  int startOfSurah = 0;
+                  String startSegmentPath = getSurahSegmentPath(audioPathInfo, start.sura,
+                      startAyahTime, endOfSurah);
+                  String lastSegmentPath = getSurahSegmentPath(audioPathInfo, end.sura,
+                      startOfSurah, endAyahTime);
+
+                  for (int surahIndex = start.sura; surahIndex <= end.sura; surahIndex++) {
+                    boolean isTheFirstSurah = surahIndex == start.sura;
+                    boolean isMiddleSurah = (surahIndex != start.sura) && (surahIndex != end.sura);
+
+                    if (isTheFirstSurah) {
+                      segmentPaths.add(startSegmentPath);
+                      continue;
+                    }
+                    if (isMiddleSurah) {
+                      segmentPaths.add(audioUtils.getSurahAudioPath(audioPathInfo, surahIndex));
+                      continue;
+                    }
+                    segmentPaths.add(lastSegmentPath);
+                  }
+
+                  boolean audioSegmentsWereCreated = !segmentPaths.isEmpty();
+
+                  if (audioSegmentsWereCreated) {
+                    audioCacheFilePaths.addAll(segmentPaths);
+                    String sharableAudioFilePath = audioUtils.getMergedAudioFromSegments(
+                        segmentPaths);
+                    shareAudioSegment(renameSharableAudioFile(sharableAudioFilePath));
+                  }
+                }
+                dismissProgressDialog();
+              }
+
+              @Override
+              public void onError(@io.reactivex.rxjava3.annotations.NonNull Throwable e) {
+                dismissProgressDialog();
+              }
+            })
+    );
+  }
+
+  private ArrayList<SparseIntArray> getTimingData(SuraAyah start, SuraAyah end,
```

Comment:
this type of thing should probably live in another module - common/audio (already exists) that we can add a timing package there and have these methods - later, the service could be updated to use this instead of replicating its code